include/deal.II/numerics/cell_data_transfer.templates.h

```diff
@@ -0,0 +1,196 @@
+        }
+      else if (cell->coarsen_flag_set())
+        {
+          // Gather the iterator to the parent cell of cells that will be
+          // coarsened. Store it together with the active cell indices of all
+          // its children.
+          Assert(cell->level() > 0, ExcInternalError());
+          const auto &parent = cell->parent();
+
+          // Check if the active_cell_indices for the current cell have
+          // been determined already.
+          if (coarsened_cells_active_index.find(parent) ==
+              coarsened_cells_active_index.end())
+            {
+              std::set<unsigned int> indices_children;
+              for (unsigned int child_index = 0;
+                   child_index < parent->n_children();
+                   ++child_index)
+                {
+                  const auto sibling = parent->child(child_index);
+                  Assert(sibling->active(), ExcInternalError());
+                  Assert(
+                    sibling->coarsen_flag_set(),
+                    ExcMessage(
+                      "Coarsen flags have not been set on all children. Please call "
+                      "Triangulation::prepare_coarsening_and_refinement() before "
+                      "calling the current function."));
+
+                  indices_children.insert(sibling->active_cell_index());
+                }
+              Assert(!indices_children.empty(), ExcInternalError());
+
+              coarsened_cells_active_index.insert({parent, indices_children});
+            }
+        }
+      else
+        {
+          // Store iterator and active cell index of all other cells.
+          persisting_cells_active_index.insert(
+            {cell, cell->active_cell_index()});
+        }
+    }
+
+#ifdef DEBUG
+  n_active_cells_pre = triangulation->n_active_cells();
+#endif
+}
+
+
+
+template <int dim, int spacedim, typename VectorType>
+void
+CellDataTransfer<dim, spacedim, VectorType>::unpack(const VectorType &in,
+                                                    VectorType &      out)
+{
+#ifdef DEBUG
+  Assert(in.size() == n_active_cells_pre,
+         ExcDimensionMismatch(in.size(), n_active_cells_pre));
+  Assert(out.size() == triangulation->n_active_cells(),
+         ExcDimensionMismatch(out.size(), triangulation->n_active_cells()));
+#endif
+
+  // Transfer data of persisting cells.
+  for (const auto &persisting : persisting_cells_active_index)
+    {
+      Assert(persisting.first->active(), ExcInternalError());
+      out[persisting.first->active_cell_index()] = in[persisting.second];
+    }
+
+  // Transfer data of the parent cell to all of its children that it has been
+  // refined to.
+  for (const auto &refined : refined_cells_active_index)
+    for (unsigned int child_index = 0;
+         child_index < refined.first->n_children();
+         ++child_index)
+      {
+        const auto child = refined.first->child(child_index);
+        Assert(child->active(), ExcInternalError());
```

Comment:
Is this always true? Couldn't I do the following:
```
+-----------+
|           |
|           |
|           |
|           |
|           |
+-----------+
```
to
```
+--+--+-----+
|__|__|     |
|  |  |     |
+--+--+-----+
|     |     |
|     |     |
+-----+-----+
```
then a refined cell could have grandchildren.

Reply:
This is indeed possible.

However, `Triangulation::prepare_coarsening_and_refinement()` cleans up all coarsening flags:  We only keep these flags whenever all siblings are active and flagged for coarsening as well. We require this specific function to be called beforehand.

This means that the exception thrown should not be `ExcInternalError`, but a custom one giving users a hint towards the `prepare()` function. Thank you for pointing that out!